### PR TITLE
ci(airflow): Replace type hints with CatalogProtocol

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

Fix https://github.com/kedro-org/kedro-plugins/issues/840

## Development notes

Replace `DataCatalog` with `CatalogProtocol`

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Review by ElenaKhaustova — COMMENTED** [review]
@ankatiyar should we also specify kedro version when release? Cause `CatalogProtocol` will be available only in `0.19.9`.

**Comment by ankatiyar** [comment]
> @ankatiyar should we also specify kedro version when release? Cause `CatalogProtocol` will be available only in `0.19.9`.

@ElenaKhaustova Yeah, you're right, let me see if I can make this work with both older and newer versions otherwise we might need to pin airflow plugin to newer versions of Kedro

**Comment by ankatiyar** [comment]
@merelcht @ElenaKhaustova I've made some changes and checked with released Kedro 0.19.8 version as well as the latest `main` version of Kedro that the tests pass and airflow plugin works. The `mypy` check doesn't for older versions complaining that `CatalogProtocol` doesn't exist in `kedro.io` but if I add an ignore statement it complains on CI because we run these tests with the `main` version so I've left it as it is. 

I've also added ignore statements to the `bandit` failures in the experimental dataset (see #850) Will leave that ticket open for the actual solution, but this will unblock the CI.

Would you mind taking another look?

**Review by merelcht — APPROVED** [review]
LGTM! 👍

File: kedro-airflow/kedro_airflow/grouping.py [hunk 1]
```diff
@@ -21,7 +30,7 @@ def get_memory_datasets(catalog: DataCatalog, pipeline: Pipeline) -> set[str]:
 
 
 def create_adjacency_list(
-    catalog: DataCatalog, pipeline: Pipeline
+    catalog: CatalogProtocol[Any] | DataCatalog, pipeline: Pipeline
```
Comment (ElenaKhaustova) [inline kedro-airflow/kedro_airflow/grouping.py:33]: qq - why do we need `CatalogProtocol[Any]` but not `CatalogProtocol`?

**Review by ElenaKhaustova — APPROVED** [review]
LGTM, thank you!

Left one non-blocking comment.

File: kedro-airflow/kedro_airflow/grouping.py (hunk 1 above)
Comment (ankatiyar) [reply kedro-airflow/kedro_airflow/grouping.py]: The original `mypy` error said something like it's getting arguments of the type `CatalogProtocol[Any]` but seems like removing this works just fine with `CatalogProtocol` so i've reverted that

**Pull request merged**